File: examples/async/aster_ws_private_readonly.py
"""Open an Aster private WebSocket user-data stream."""

import asyncio
import os

from dcex.ws import aster


async def main() -> None:
    """Open the user-data stream, keep it alive, and print one message."""
    user_address = os.environ["ASTER_USER_ADDRESS"]
    signer_address = os.environ["ASTER_SIGNER_ADDRESS"]
    private_key = os.environ["ASTER_PRIVATE_KEY"]

    async with aster.private(
        user_address=user_address,
        signer_address=signer_address,
        private_key=private_key,
        market="futures",
    ) as ws:
        await ws.keep_alive()
        print(await ws.recv())


if __name__ == "__main__":
    asyncio.run(main())

File: crates/dcex/examples/aster_ws_private_readonly.rs
use std::env;
use std::time::Duration;

use dcex::exchanges::aster::AsterMarket;
use dcex::ws::aster::AsterPrivateWebSocket;

#[tokio::main]
async fn main() -> dcex::Result<()> {
    let user_address = env::var("ASTER_USER_ADDRESS")
        .map_err(|error| dcex::DcexError::InvalidInput(error.to_string()))?;
    let signer_address = env::var("ASTER_SIGNER_ADDRESS")
        .map_err(|error| dcex::DcexError::InvalidInput(error.to_string()))?;
    let private_key = env::var("ASTER_PRIVATE_KEY")
        .map_err(|error| dcex::DcexError::InvalidInput(error.to_string()))?;

    let mut ws = AsterPrivateWebSocket::new(
        Some(user_address),
        signer_address,
        private_key,
        AsterMarket::Futures,
        Duration::from_secs(10),
    )?;
    ws.connect().await?;
    ws.keep_alive().await?;
    println!("{}", ws.recv().await?);
    ws.close().await?;
    Ok(())
}
